### PR TITLE
Better error reporting for union decoders

### PR DESCRIPTION
**Pull request by tomrf1**
The union decoder was previously swallowing the reason for failure in the nested object. It was turning the decoding result (an Either) into an Option.
This made debugging the capi issue harder.

### CODE REVIEW

File: src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala [hunk 1]
```diff
@@ -238,17 +238,17 @@ private class CirceScroogeMacrosImpl(val c: blackbox.Context) {
     q"""{
       new _root_.io.circe.Decoder[$A] {
         def apply(c: _root_.io.circe.HCursor): _root_.io.circe.Decoder.Result[$A] = {
-          val result = c.fields.getOrElse(Nil).headOption.flatMap {
+          val maybeResult = c.fields.getOrElse(Vector.empty).headOption.flatMap {
```
Comment (mchv) [inline src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala]: encoding the type in the variable name [`hungarian notation`](https://en.wikipedia.org/wiki/Hungarian_notation) is usually not recommended in scala

Comment (mchv) [reply src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala]: why `Vector.empty` instead of `Nil`, for performance reason?

Comment (tomrf1) [reply src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala]: `c.fields` returns a `Option[Vector]` (used to be a List). Changing it from Nil makes no difference as far as I know, as `headOption` is common to both, but it feels more correct.

Comment (mchv) [reply src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala]: 👍

**Comment by mchv** [comment]
💯

File: src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala (hunk 1 above)
Comment (LATaylor-guardian) [reply src/main/scala/com/gu/fezziwig/CirceScroogeMacros.scala]: > encoding the type in the variable name hungarian notation is usually not recommended in scala

I personally find `maybeResult` for an optional value much better for readability than just result. The alternative if not preferred it to always explicitly declare the type.

**Pull request merged**